build: Find feature headers outside system include dir

camscale opened the pull request:
Look for PAM and BPF header files that control including features in
teleport binaries more generally instead of assuming they are in the
system include headers or the one location in the buildbox (for libbpf
in `/usr/libbpf-1.2.2`).

This allows a cross-compiler to be used that sets the standard
environment variables `C_INCLUDE_PATH` and `LIBRARY_PATH` to point to
where additional headers and libraries live.

This cleans up the BPF detection in `common.mk` as it had some
head-scratching failure modes when clang or llvm-strip were not found.
The logic for enabling BPF is clearer now and safer.

Make `IS_NATIVE_BUILD` into a proper make variable - empty means false,
non-empty means true, and adjust the Makefile where it is used.

camscale committed:
build: Use pkg-config to find bpf headers and libraries

Use `pkg-config` / `pkgconf` to find the specific version of `libbpf`
that we want to link to as well as the dependent libraries such as
`libelf`, `libz`, and sometimes `libzstd`. This resolves the variable
dependency on `libzstd` depending on the base OS.

We still take preference of `/lib/libbpf-$(LIBBPF_VER)` if it exists as
it does for the standard buildbox, but that is deprecated and will be
removed when we move to the new buildbox.

If there is no `libelf.pc` config, fall back to hard-coded libraries as
it was previously. Again, this is for use with our existing CentOS 7
buildbox.

This cleans up the BPF detection in `common.mk` as it had some
head-scratching failure modes when clang or llvm-strip were not found.
The logic for enabling BPF is clearer now and safer.

camscale commented:
Test builds:
* amd64: https://github.com/gravitational/teleport.e/actions/runs/9854414086
* amd64-fips: https://github.com/gravitational/teleport.e/actions/runs/9854414556
* arm64: https://github.com/gravitational/teleport.e/actions/runs/9854959439
* 386: https://github.com/gravitational/teleport.e/actions/runs/9854415243
* arm: https://github.com/gravitational/teleport.e/actions/runs/9854959818

r0mant commented on Makefile:
```diff
@@ -71,18 +71,16 @@ FIPS_MESSAGE := with-FIPS-support
 endif
 
-# PAM support will only be built into Teleport if headers exist at build time.
+# Look for the PAM header "security/pam_appl.h" to dertermine if we should
```
```suggestion
# Look for the PAM header "security/pam_appl.h" to determine if we should
```

camscale commented:
Jakub has made some changes that conflicts with this that I need to consider how it should integrate. I'll rework this shortly.

jakule commented on common.mk:
```diff
@@ -40,11 +65,34 @@ ER_BPF_BUILDDIR := lib/bpf/bytecode
+STATIC_LIBS += $(addprefix -L,$(subst :, ,$(LIBRARY_PATH)))
+
+# Libraries needed for or by BPF
+STATIC_LIBS += -lbpf -lelf -lz
```
FYI, those are not the correct libraries all the time. libelf added in newer versions dependency on libzstd, so I used pkg-conf to get the "correct" version. Otherwise I'm using the same set to not break build on systems without pkg-conf.

As I mentioned a few times IMO we should merge https://github.com/gravitational/teleport/pull/36106 and stop dealing with libbpf and all related issues.
CC @r0mant

camscale replied:
I've added a libelf.pc file to buildbox-ng so I should be able to use your refactoring here and not hard-code these libraries unless we do not find a libelf.pc at all.

camscale commented:
@jakule I've incorporated your libelf pkg-config changes and also started using pkg-config for libbpf too, pinning to the specific version we require. I'm not sure if I got a little silly using `false` as the fallback for `pkg-config` when neither it or `pkgconf` are installed. It seems to work well enough, but is perhaps a little tricky? let me know.

This still conflicts because I need to rebase, but the contents of this PR is what the final version would look like if you just view the full file (common.mk).

jakule commented on common.mk:
```diff
@@ -1,33 +1,81 @@
+# -----------------------------------------------------------------------------
+# libbpf detection
+#
+# Requirements for building with BFP support:
```
```suggestion
# Requirements for building with BPF support:
```

jakule commented on common.mk:
```diff
@@ -1,33 +1,81 @@
+else
+ifneq (,$(shell $(PKGCONF) --exists 'libbpf = $(LIBBPF_VER)' && echo true))
+FOUND_LIBBPF := true
+LIBBPF_INCLUDES := $(shell $(PKGCONF) --cflags libbpf)
+LIBBPF_LIBS := $(shell $(PKGCONF) --libs --static libbpf)
```
This will break our OSS build for some users that have libbpf installed in the system (unless now we only provide the official builds).
The problem with our Go libbpf wrapper is that not backward compatible, so you MUST have the expected version. `pkg-conf` returns flags if any version is installed. 
Normally I'd use pkg-conf for everything, but in this case I don't think we should be using it. There is very small change that your OS have the correct version installed and flags returned by pkg-conf are always related to the system wide installed version.

camscale replied:
I thought I addressed this a couple of lines above with:
```
ifneq (,$(shell $(PKGCONF) --exists 'libbpf = $(LIBBPF_VER)' && echo true))
```

This block is only executed if the specific version of libbpf is installed with pkgconf. Do you think there can be a problem if multiple versions of libbpf are installed? The man page for pkg-config suggests that using something like `--exact-version=X.Y.Z` can only be used with `--exists`, and even then its unnecessary due to `--exists` supporting version constraints as I've used. Perhaps I can use `--lib` with `--exact-version` to ensure we get the correct libs - I'll test that out.

jakule replied:
Aaaa, sorry. I missed this part. TBH I don't have strong opinion about it. Feel free to merge it.

camscale committed:
build: Also look in C_INCLUDE_PATH for pam_appl.h

When looking for the PAM header `pam_appl.h` to determine if PAM should
be enabled in teleport, first look in the directories specified in the
`C_INCLUDE_PATH` environment variable before checking the system header
directories. The buildbox-ng has per-architecture include directories
and configures `C_INCLUDE_PATH` to find the correct headers. Fall back
to the existing default directories if not found in `C_INCLUDE_PATH`.

camscale commented:
I've updated this PR to squash the commits together as that was the easiest way to resolve the conflicts, which I then rebased with `--strategy-option=theirs`. I've done a local test build with the centos7 buildbox and it is working as expected. I have decided to keep the pkg-config setup for libbpf as it is working as expected and is a little more general.

I'll merge this tomorrow morning my time in case there are any further objections or comments.
